File: app/shared/work_details.py
"""Parse vehicle work details embedded in field-service comments."""

from __future__ import annotations

import json
from dataclasses import dataclass
from decimal import Decimal

_MARKER_START = "[kf:work]"
_MARKER_END = "[/kf:work]"


@dataclass(frozen=True)
class VehicleWorkDetails:
    profile: str
    crop_code: str | None = None
    crop_name: str | None = None
    area_acres: str | None = None
    cultivation_stage: str | None = None
    trips: str | None = None
    purpose: str | None = None
    material: str | None = None
    locality: str | None = None
    distance_km: str | None = None
    weight_kg: str | None = None
    goods_type: str | None = None
    tonnes: str | None = None
    loading_point: str | None = None
    unloading_point: str | None = None
    litres: str | None = None
    spray_type: str | None = None
    bale_count: str | None = None


def parse_work_details_from_comments(comments: str | None) -> tuple[VehicleWorkDetails | None, str]:
    if not comments:
        return None, ""
    start = comments.find(_MARKER_START)
    end = comments.find(_MARKER_END)
    if start == -1 or end == -1 or end <= start:
        return None, comments
    raw = comments[start + len(_MARKER_START) : end].strip()
    free = f"{comments[:start]}{comments[end + len(_MARKER_END) :]}".strip()
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        return None, comments
    if not isinstance(data, dict) or not data.get("profile"):
        return None, comments
    return VehicleWorkDetails(**{k: data.get(k) for k in VehicleWorkDetails.__dataclass_fields__}), free


def _to_int(value: str | int | None) -> int | None:
    if value is None or value == "":
        return None
    try:
        return int(Decimal(str(value)))
    except Exception:
        return None


def trips_from_work_details(details: VehicleWorkDetails | None) -> int | None:
    if details is None or not details.trips:
        return None
    return _to_int(details.trips)


def bale_count_from_work_details(details: VehicleWorkDetails | None) -> int | None:
    if details is None:
        return None
    if details.bale_count:
        return _to_int(details.bale_count)
    if details.profile == "baler" and details.trips:
        return _to_int(details.trips)
    return None


def strip_work_details_marker(comments: str | None) -> str | None:
    _, free = parse_work_details_from_comments(comments)
    return free or None

File: frontend/src/features/field-services/work-details.ts
/** Vehicle-type-specific work details stored in field-service comments. */

export const CULTIVATION_STAGE_OPTIONS = [
  { value: "land_preparation", label: "Land preparation" },
  { value: "sowing", label: "Sowing" },
  { value: "intercultivation", label: "Intercultivation" },
  { value: "weeding", label: "Weeding" },
  { value: "harvesting", label: "Harvesting" },
  { value: "other", label: "Other" },
] as const;

export const TROLLEY_PURPOSE_OPTIONS = [
  { value: "grain_haul", label: "Grain haul" },
  { value: "fertilizer", label: "Fertilizer" },
  { value: "seed", label: "Seed" },
  { value: "soil", label: "Soil" },
  { value: "other", label: "Other" },
] as const;

export const TROLLEY_MATERIAL_OPTIONS = [
  { value: "paddy", label: "Paddy" },
  { value: "corn", label: "Corn" },
  { value: "fertilizer", label: "Fertilizer" },
  { value: "soil", label: "Soil" },
  { value: "bags", label: "Bags" },
  { value: "other", label: "Other" },
] as const;

export const LOCALITY_OPTIONS = [
  { value: "local", label: "Local" },
  { value: "non_local", label: "Non-local" },
] as const;

export const GOODS_TYPE_OPTIONS = [
  { value: "grain", label: "Grain" },
  { value: "fertilizer", label: "Fertilizer" },
  { value: "seeds", label: "Seeds" },
  { value: "tools", label: "Tools" },
  { value: "other", label: "Other" },
] as const;

export const SPRAY_TYPE_OPTIONS = [
  { value: "pesticide", label: "Pesticide" },
  { value: "fertilizer", label: "Fertilizer" },
  { value: "growth_regulator", label: "Growth regulator" },
  { value: "other", label: "Other" },
] as const;

export interface VehicleWorkDetails {
  profile: "tractor" | "trolley" | "bolero" | "dcm" | "pump" | "drone" | "baler";
  // Tractor / implements / harvester
  crop_code?: string;
  crop_name?: string;
  area_acres?: string;
  cultivation_stage?: string;
  // Trolley / baler
  trips?: string;
  bale_count?: string;
  purpose?: string;
  material?: string;
  // Bolero
  locality?: string;
  distance_km?: string;
  weight_kg?: string;
  goods_type?: string;
  // DCM
  tonnes?: string;
  loading_point?: string;
  unloading_point?: string;
  // Pump
  litres?: string;
  // Drone
  spray_type?: string;
}

const MARKER_START = "[kf:work]";
const MARKER_END = "[/kf:work]";

export function parseWorkDetailsFromComments(comments: string | null | undefined): {
  details: VehicleWorkDetails | null;
  freeComments: string;
} {
  if (!comments) return { details: null, freeComments: "" };
  const start = comments.indexOf(MARKER_START);
  const end = comments.indexOf(MARKER_END);
  if (start === -1 || end === -1 || end <= start) {
    return { details: null, freeComments: comments };
  }
  const json = comments.slice(start + MARKER_START.length, end).trim();
  const freeComments = `${comments.slice(0, start)}${comments.slice(end + MARKER_END.length)}`.trim();
  try {
    const details = JSON.parse(json) as VehicleWorkDetails;
    if (!details?.profile) return { details: null, freeComments: comments };
    return { details, freeComments };
  } catch {
    return { details: null, freeComments: comments };
  }
}

export function mergeWorkDetailsIntoComments(
  freeComments: string,
  details: VehicleWorkDetails | null,
): string | null {
  const free = freeComments.trim();
  if (!details) return free || null;
  const block = `${MARKER_START}${JSON.stringify(details)}${MARKER_END}`;
  return free ? `${block}\n${free}` : block;
}

/** Touch-friendly MUI TextField defaults for field ops forms. */
export const TOUCH_FIELD_SX = {
  "& .MuiInputBase-root": { minHeight: 52 },
  "& .MuiSelect-select": { py: 1.5 },
} as const;
